### ISSUE TITLE
Define user account model with web2 login support

### ISSUE THREAD

**Issue by ryanchristo** [regen-network/regen-server#375]
## Summary

With the addition of web2 authentication (https://github.com/regen-network/regen-web/issues/2071), we are re-evaluating our user account model.

In the current implementation, an `account` can be linked to one or more `party`. Each `party` the account is linked to represents an individual or an organization, i.e. each `party` represents a single identity (i.e. an `account` can have multiple identities).

~If the `party` represents an individual, the party is *not* linked to an `organization`. If the `party` represents an organization, the party *is* linked to an `organization`.~ Each `party` is linked to a `wallet`, therefore each wallet also represents an identity.

Questions:
- If the user logs in with web2 authentication (no wallet), how do we link a `party` to the user?
- Should a single `account` be linked to one or more `party` and therefore one or more `wallet` as we have currently?

## Diagram (Current)
 
![Screenshot from 2023-09-18 12-24-26](https://github.com/regen-network/regen-server/assets/12519942/2a3872a1-499d-40fa-8148-58639d8671dc)

**Comment by wgwz** [comment]
i don’t think we actually really use the organization table for anything important. as part of the db clean up task we did months ago this was identified as something that we could get rid of in a follow up task. at the time we did the clean up the only place the organization table was being used was in the roles form during project create flow. that may not even be true anymore cc. @blushi 

i think we can remove that as a consideration for this discussion though, but we should still think about that as a clean up task if it hasn’t been handled already. it doesn’t have any effect on the login system otherwise

**Comment by wgwz** [comment]
the party.type column is actually what determines what type of party (profile) it is, it’s value is one of two things, “user” or “organization”.

**Comment by ryanchristo** [comment]
Thinking about this alongside #376, I would propose we update `account` to `session` and `party` to `account`. As far as I can tell, "session" is the same terminology used in social media platforms when the user has the option to add multiple accounts and switch between them and this is essentially how we are using `account` at the moment. An "account" is usually associated with a user identity, i.e. it has a name, email, username, wallet address, etc. which is how we are using `party` at the moment.

This also aligns with how lucia is using sessions: https://lucia-auth.com/basics/sessions/ where a user can have multiple active sessions (which are managed remotely while the session id is stored in a cookie or token locally). Lucia uses `user` where I am proposing `account` in their example documentation, which could be another option for `party` (instead of `account`).

With the addition of email, and following our current architecture, we could mimic `wallet` and add an `email` table and we would add `email_id` to `account` to link an email. If we add multiple social login options, we could do the same for each, for example `google_id` would link to a `google` table storing social login information for a user authenticated with `google`.

Lucia examples add these directly to the `user` (or `account`) table (example uses `username`) but there may be reasons to use our existing approach such as having an email that is stored but uncomfirmed.

A user would still be able to switch between accounts, as they currently can switch between "addresses" (we need to update the language used here in the application as well). Each account would be associated with a single profile, and at least initially, with a single wallet address and a single email address (and eventually a single social account per platform).

**Comment by ryanchristo** [comment]
Adding a note with regards to authentication, a user would have multiple authentication options for logging into an account, i.e. using arbitrary message signing with keplr (if wallet connected to account), using email (if email connected to account), or using social login (if social connected to account). Each login option would create an active session for the one account.

**Comment by ryanchristo** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/RND-Dev-Team-5f8998bec8958d000f4609e2/issues/regen-network/regen-server/375?planning-poker) with Zenhub @blushi @flagrede @ryanchristo @wgwz

**Comment by blushi** [comment]
> i don’t think we actually really use the organization table for anything important. as part of the db clean up task we did months ago this was identified as something that we could get rid of in a follow up task. at the time we did the clean up the only place the organization table was being used was in the roles form during project create flow. that may not even be true anymore cc. @blushi
> 
> i think we can remove that as a consideration for this discussion though, but we should still think about that as a clean up task if it hasn’t been handled already. it doesn’t have any effect on the login system otherwise

indeed we should probably remove the organization table now it's now being used anymore as part of the project create flow.

**Comment by blushi** [comment]
we do have a concept of organization already, in the current `party` table, we have a `type` column which is either "user" or "organization" (the reason for having an organization table was that we also had a `user` table, and for a given `party` there would be one `user` or `organization` associated to it to store info only relevant to `user` or `organization`).

**Comment by wgwz** [comment]
based on what i've heard in the recent discussions, i'd like to propose the following changes:

1. removal of the `account` table as it stands today
2. renaming of the `party` table to the `account` table
3. removal of the `wallet` table
4. add `addr` and `email` column to the new `account` table

more on 3.:

this would potentially be a nice simplification in light of the latest desired functionality in the login system.
specifically how one account:
- can have only one wallet address
- can have only one email address
- can have only one connected twitter, linkedin or google account
it adds uniformity to how the database model works.
currently the `party` to `wallet` relationship is one-to-one anyways.
so why not just collapse the relationship into the `party` table?
we only use wallet to store an address anyways.
but no strong opinion.

more on 4.:

it seems like regardless of the social login provider, we could always key off of the email address, assuming that each platform can be configured to return the email address of the user during oauth.
that's potentially nice because then we only ever have to store an email in the new `account` table.
and we would not for example have to store a different type of external identifier related to the users social account.
regardless of the social login method, we always are just matching the email returned by the social login provider with what we have in our database.

more on all of the above:

i took some time to point out various locations affected by the suggested changes above:
- https://github.com/regen-network/regen-server/pull/384

in summary:
- we can simplify the current relevant db schema where `account` to `party` to `wallet` to a single table called `account`.
- alternatively `account` could be our `users` table.

sessions:
- we can just set up a standard, stand-alone session table as it's usually done via express-session and database session adapters: https://github.com/voxpelli/node-connect-pg-simple#installation

**Comment by blushi** [comment]
@ryanchristo @wgwz could you add your estimate for this? thanks

**Comment by blushi** [comment]
An edge case that would make the current `wallet` table still useful is the following:
- A user logs in with email/social initially then attach a keplr wallet address to the account later on
- The user creates some on-chain projects
- Then the user decides to remove the wallet address from the account (although I know this is not something we would want to support initially...)
- I don't think we would want such user to be authorized to update the unanchored metadata for these projects since the account is not linked to the projects admin wallet address anymore. Currently, we have a `admin_wallet_id` column in the `project` table on which we base project updates authorization, which would handle this particular edge case. But if we decided to remove the `wallet` table and replace `admin_wallet_id` with `admin_party/account_id`, this wouldn't handle it very well.

**Comment by blushi** [comment]
@wgwz should we close this now?
